In [1]:
import numpy as np

In [3]:
#Se define la función
def f(x): return np.exp(-x**2)

#Se estima el máximo de la segunda derivada en el intervalo [0,1] con un h de 0.001
#Note que el orden de la estimación sería de h^2 = 10^-6
x, h = np.linspace(0, 1, 1000, retstep=True)
d = (f(x+h)-2*f(x)+f(x-h))/h**2
maxd = np.max(np.abs(d))

#Se imprime, teniendo en cuenta el orden de la estimación, el valor máximo aproximado de la segunda derivada
str(maxd)[0:7]

'1.99999'

A partir de la fórmula del error del método del trapecio compuesto $|E|\leq \frac{h^2(b-a)}{12}max|f^{''}(\xi)|$, se obtiene la expresión siguiente:
$$n \geq \sqrt{\frac{(b-a)^3}{12E}max|f^{''}(\xi)|}$$

Con dicha expresión se calcula el valor mínimo de $n$ a partir del valor $E=0.005$, dado en el enunciado, y del valor $max|f^{''}(\xi) = 1.99999$ obtenido anteriormente:
$$n \geq \sqrt{\frac{(1)^3}{12(0.005)}(1.99999)} \approx 5.77349 = 6$$

Por lo tanto, se prosigue a calcular la integral utilizando el método del trapecio compuesto con $n = 6$ trapecios, es decir, $n + 1 = 7$ puntos.

In [9]:
def uniform_trapezoid(y, h):
    integral = 0.
    integral += 0.5*(y[0]+y[-1])
    integral += np.sum(y[1:-1])

    return integral*h

def uniform_trapezoid_error(x, f, h):
    d = (f(x+h)-2*f(x)+f(x-h))/h**2
    maxd = np.max(np.abs(d))

    error = (maxd* (x[-1]-x[0])**3 )/(12*(len(x)-1)**2)

    return error

n = 6
x = np.linspace(0, 1, n+1)
y = f(x)
h = (x[-1]-x[0])/n

integral = uniform_trapezoid(y, h)
error = uniform_trapezoid_error(x, f, h)

"Integral: " + str(integral), "Error: " + str(error)

('Integral: 0.7451194124361793', 'Error: 0.004565920480608619')

Consecuentemente, el resultado sería:
$$\int_{0}^{1}e^{-x^2}dx \approx (0.745 \pm 0.005)$$

Sin embargo, como la incertidumbre queda igual a 0.005 con el redondeo usual a una cifra significativa, entonces se utiliza $n = 7$ trapecios para calcular la integral y obtener un error estrictamente menor a 0.005 tras redondear la incertidumbre.

In [10]:
n = 7
x = np.linspace(0, 1, n+1)
y = f(x)
h = (x[-1]-x[0])/n

integral = uniform_trapezoid(y, h)
error = uniform_trapezoid_error(x, f, h)

"Integral: " + str(integral), "Error: " + str(error)

('Integral: 0.7455719918300938', 'Error: 0.0033668876910492775')

Luego el resultado final es:
$$\int_{0}^{1}e^{-x^2}dx \approx (0.746 \pm 0.003)$$